internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 	"github.com/filecoin-project/go-filecoin/internal/pkg/vm/address"
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
```

Comment:
Any error encountered by the subcall (either `rt.Fatal`, `panic`, or an ordinary error return code) cascades and immediately causes the caller to also fail with `MethodSubcallError`, unless the caller explicitly uses `SendAllowingErrors`. This is primarily to prioritize "fail safe" behavior -- i.e., if actor A calls actor B and doesn't realize the subcall might fail, we would rather the entire transaction fail to apply, than for actor A to continue erroneously and complete the transaction.

(Related question, though -- if naming the error-cascading version `Send` is too error-prone, we could rename it to `SendCascadingErrors` or something similar, and not have an unspecified general `Send` method.)

Reply:
This is an implementation detail and the mechanism to make that happen is very language specific.

If you are peusocoding the actors in "Go-ish" then ignore the error in go style, or give your pseudocode language a `try/catch` and just do:

```
try
   runtime.Send(blah)
catch
    // ignore 
end
```

It cleans up the interface and lets the implementors decide how to build the VM primitives the way they want and in a way that is closer to the language it is in. 

Btw, what actor needs to do calls on other actors and ignore errors? And why would we want actors to be able to call other actors that fail and carry on like nothing happen? Do we expect to reach the same state as other instances that did not see the error? How does that even work? An actor providing a pure function that fails?

Reply:
> Btw, what actor needs to do calls on other actors and ignore errors? And why would we want actors to be able to call other actors that fail and carry on like nothing happen? Do we expect to reach the same state as other instances that did not see the error? How does that even work? An actor providing a pure function that fails?

Errors that are handled in this fashion must be actual VM/actor errors, not nondeterministic failures (see below). For the former, one example is the cron actor -- suppose actors A, B, C, ... register with cron; then it should not be possible for A to DoS the cron calls to B and C by raising an error within A itself.

Reply:
As currently written in the spec, that `SendAllowingErrors` would allow the caller to trap an "abort()" and also an unhandled `panic()`, is that intended?

Reply:
No -- the intent is for `SendAllowingErrors` (N.B.: renamed to `SendCatchingErrors`) to only intercept actual VM/actor errors, not nondeterministic failures or unhandled panics. Will draft a PR to change this in the spec.